src/core/rpc_client.cc

```diff
@@ -167,34 +175,67 @@ process_body_for_self(stdx::optional<rpc_recv_context> opt, auto self) {
+    if (SMF_UNLIKELY(!hdr)) {
+      if (!rpc_slots.empty()) {
+        LOG_ERROR("Could not parse response from server. Bad Header");
+        conn->set_error("Could not parse header from server");
+      } else {
+        // expected. no outstanding reqs
+        conn->disable();
+      }
+      return seastar::make_ready_future<>();
+    }
+    return rpc_recv_context::parse_payload(conn.get(), std::move(hdr.value()))
+      .then([this](stdx::optional<rpc_recv_context> opt) mutable {
+        if (!opt) {
+          LOG_ERROR("Could not parse response from server. Bad payload");
+          conn->set_error("Invalid payload");
+        } else if (read_counter > 0 && opt) {
+          --read_counter;
+          auto v = std::move(opt.value());
+          auto sess = v.session();
+          auto it = rpc_slots.find(sess);
+          if (it != rpc_slots.end()) {
+            auto slot = it->second;
+            rpc_slots.erase(sess);
+            DLOG_THROW_IF(slot->session != v.session(),
+                          "Invalid client sessions");
+            using typed_ret = decltype(opt);
+            slot->pr.set_value(typed_ret(std::move(v)));
+          }
+        }
+        return seastar::make_ready_future<>();
+      });
+  });
+}
 seastar::future<>
 rpc_client::do_reads() {
+  auto self = parent_shared_from_this();
   return seastar::do_until(
            [c = conn] { return !c->is_valid(); },
-           [self = parent_shared_from_this()]() mutable {
-             return rpc_recv_context::parse_header(self->conn.get())
-               .then([self](auto hdr) {
-                 if (SMF_UNLIKELY(!hdr)) {
-                   if (!self->rpc_slots.empty()) {
-                     LOG_ERROR(
-                       "Could not parse response from server. Bad Header");
-                     self->conn->set_error(
-                       "Could not parse header from server");
-                   } else {
-                     // expected. no outstanding reqs
-                     self->conn->disable();
-                   }
-                   return seastar::make_ready_future<>();
-                 }
-                 return rpc_recv_context::parse_payload(self->conn.get(),
-                                                        std::move(hdr.value()))
-                   .then([self](stdx::optional<rpc_recv_context> opt) mutable {
-                     return process_body_for_self(std::move(opt), self);
-                   });
+           [this]() {
+             auto timeout = conn->limits->max_body_parsing_duration;
+             seastar::timer<> body_timeout;
+             body_timeout.set_callback([timeout, this] {
+               LOG_ERROR(
+                 "rpc_client exceeded timeout: {}ms",
+                 std::chrono::duration_cast<std::chrono::milliseconds>(timeout)
+                   .count());
+               conn->set_error("Connection body parsing exceeded timeout");
```

Comment:
should instead `fullfill_timeoutsession(uint16_t) which can just pass in nullopt ? `

need to brainstorm.